prov/usnic/src/usdf_dgram.c

```diff
@@ -209,112 +212,134 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
 	return 0;
 }
 
+static ssize_t
+_usdf_dgram_send_iov(struct usdf_ep *ep, const struct usd_dest *dest,
+	const struct iovec *iov, size_t count, void *context,
+	uint8_t cq_entry)
+{
+	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
+
+	struct usd_wq_post_info *info;
+	struct usd_udp_hdr *hdr;
+	struct usd_qp_impl *qp;
+	struct usd_wq *wq;
+
+	uint32_t last_post;
+	uint8_t *copybuf;
+	size_t len = 0;
+
+	for (size_t i = 0; i < count; i++) {
+		len += iov[i].iov_len;
+	}
+
+	qp = to_qpi(ep->e.dg.ep_qp);
+	wq = &qp->uq_wq;
+	copybuf = wq->uwq_copybuf +
+			(wq->uwq_post_index * USD_SEND_MAX_COPY);
+	hdr = (struct usd_udp_hdr *)copybuf;
+	memcpy(hdr, &dest->ds_dest.ds_udp.u_hdr, sizeof(*hdr));
+
+	/* adjust lengths and insert source port */
+	hdr->uh_ip.tot_len = htons(len + sizeof(struct usd_udp_hdr) -
+		sizeof(struct ether_header));
+	hdr->uh_udp.len = htons(len + sizeof(struct usd_udp_hdr) -
+		sizeof(struct ether_header) -
+		sizeof(struct iphdr));
+	hdr->uh_udp.source =
+		qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
+
+	send_iov[0].iov_base = hdr;
+	send_iov[0].iov_len = sizeof(*hdr);
+	memcpy(&send_iov[1], iov, sizeof(struct iovec) * count);
+	last_post = _usd_post_send_iov(wq, send_iov, count + 1,
+		cq_entry);
+	info = &wq->uwq_post_info[last_post];
+	info->wp_context = context;
+	info->wp_len = len;
+
+	return 0;
+}
+
 ssize_t
 usdf_dgram_sendv(struct fid_ep *fep, const struct iovec *iov, void **desc,
 		 size_t count, fi_addr_t dest_addr, void *context)
 {
 	struct usdf_ep *ep;
 	struct usd_dest *dest;
-	struct usd_wq *wq;
- 	struct usd_qp_impl *qp;
-	struct usd_udp_hdr *hdr;
-	uint32_t last_post;
-	struct usd_wq_post_info *info;
-	uint8_t *copybuf;
-	size_t len;
-	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
-	size_t i;
+	uint8_t completion;
+	size_t send_size = sizeof(struct usd_udp_hdr);
```

Comment:
The style in the rest of the code is generally (though probably not 100% consistently) to only declare variables and then initialize them with later statements.  My preference is to continue that style unless there's a good reason to do otherwise.